src/cryptoadvance/specter/services/swan/controller.py

```diff
@@ -0,0 +1,220 @@
+    destination_wallet_alias = request.form["destination_wallet"]
+    wallet = current_user.wallet_manager.get_by_alias(destination_wallet_alias)
+
+    try:
+        SwanService.set_autowithdrawal_settings(wallet=wallet, btc_threshold=threshold)
+        return redirect(url_for(f"{SwanService.get_blueprint_name()}.withdrawals"))
+    except swan_api.SwanServiceApiException as e:
+        logger.exception(e)
+        flash(
+            _("Error communicating with Swan API")
+        )
+        return redirect(url_for(f"{SwanService.get_blueprint_name()}.settings"))
+
+
+
+
+""" ***************************************************************************
+                                OAuth2 endpoints
+*************************************************************************** """
+@swan_endpoint.route("/oauth2/start")
+@swan_endpoint.route("/oauth2/start/<is_relink>")
+@login_required
+@user_secret_decrypted_required
+def oauth2_start(is_relink=0):
+    """
+    Set up the Swan API integration by requesting our initial access_token and
+    refresh_token.
+    """
+    # Do we have a token already?
+    if SwanService.is_access_token_valid():
+        return redirect(url_for(f"{SwanService.get_blueprint_name()}.settings"))
+
+    # Let's start the PKCE-flow
+    flow_url = swan_api.get_oauth2_start_url()
+
+    print(f"current_user: {current_user}")
+    print(f"is_relink: {is_relink}")
+    return render_template(
+        "swan/oauth2_start.jinja",
+        flow_url=flow_url,
+        is_relink=int(is_relink) == 1,
+    )
+
+
+"""
+    Note: the callback from Swan will be treated by Flask as an AnonymousUserMixin request
+    but we need the user logged in and their user_secret decrypted. So we must require
+    @login_required here which will interrupt the return flow with the login prompt.
+"""
+@swan_endpoint.route("/oauth2/callback")
+@login_required
+def oauth2_auth():
+    if request.args.get("error"):
+        logger.error(
+            f"OAuth2 flow error: {request.args.get('error')}, {request.args.get('error_description')}"
+        )
+        return render_template(
+            "500.jinja",
+            error=request.args.get("error"),
+            traceback=request.args.get(
+                "error_description"
+            ),  # Slightly misusing the traceback field
+        )
+    
+    user = app.specter.user_manager.get_user()
+
+    try:
+        swan_api.handle_oauth2_auth_callback(request)
+
+        # Add the Service to the User's profile (will now appear in sidebar)
+        user = app.specter.user_manager.get_user()
+        if SwanService.id not in user.services:
+            user.services.append(SwanService.id)
+
+        # Sync Specter with any previous Swan-Specter integrations
+        # TODO: This still needs more testing; it's irrelevant for first-time Swan authorizations anyway
+        # SwanService.sync_swan_data()
+
+        return redirect(url_for(".oauth2_success"))
+    except Exception as e:
```

Comment:
This particular one is encapsulating a lot. It's doing the api callback to exchange the `authorization_code` for the `access_token` and then calling the complex `sync_swan_data` (which still needs a lot of testing). Would probably make sense to at least break it up into two separate `try/except` blocks rather than stacking `Exception` handlers.